File: web_version/pmugg_files/shapes3D/bspPolygon.js
ms.bspPolygon = function(points, face) {
	var stats = face.getNode().getStats();
	var properties = {
		bspNode: new ms.singleProperty('bspNode', false),
		face: new ms.singleProperty('face', true),
	};
	this.node = new ms.node(this, stats, 'bspPolygon', properties);
	this.points = points;

	face.getNode().connect(this);
};

ms.bspPolygon.prototype.getNode = function() {
	return this.node;
};

ms.bspPolygon.prototype.getFace = function() {
	return this.node.get('face')
};

ms.bspPolygon.create = function(face) {
	var points = face.getEndpoints().map(function(endpoint) {
		return endpoint.getPosition();
	});
	return new ms.bspPolygon(points, face);
};

ms.bspPolygon.prototype.getPoints = function() {
	return this.points;
};

ms.bspPolygon.prototype.selfIntersects = function() {
	for (var i = 0; i < this.points.length - 1; i++) {
		var a0 = this.points[i];
		var a1 = this.points[i + 1];
		for (var j = i + 1; j < this.points.length - 1; j++) {			
			var b0 = this.points[j];
			var b1 = this.points[j + 1];
			if (ms.intersector.intersect3D(a0, a1, b0, b1)) {
				return true;
			}
		}
	}
	return false;
};

ms.bspPolygon.prototype.print = function() {
	this.getFace().print();
};

File: cpp_version/pmugg/shapes3D/bsp_polygon.cpp
#include "bsp_polygon.h"
#include "face.h"
#include "node.h"
#include "single_property.h"
#include "intersector.h"
#include "endpoint.h"

namespace ms {

BspPolygon::BspPolygon(const std::vector<Vec3>& pts, Face* f)
    : points(pts) {
    auto stats = f->getNode()->getStats();
    
    properties.bspNode = std::make_unique<SingleProperty>("bspNode", false);
    properties.face = std::make_unique<SingleProperty>("face", true);
    
    node = std::make_unique<Node>(this, stats, "bspPolygon", properties);
    
    f->getNode()->connect(this);
}

Face* BspPolygon::getFace() const {
    return node->get("face");
}

bool BspPolygon::selfIntersects() const {
    for (size_t i = 0; i < points.size() - 1; i++) {
        const Vec3& a0 = points[i];
        const Vec3& a1 = points[i + 1];
        
        for (size_t j = i + 1; j < points.size() - 1; j++) {
            const Vec3& b0 = points[j];
            const Vec3& b1 = points[j + 1];
            
            if (Intersector::intersect3D(a0, a1, b0, b1)) {
                return true;
            }
        }
    }
    return false;
}

void BspPolygon::print() const {
    getFace()->print();
}

} // namespace ms 